### DISCUSSION TITLE
Tasmota on ESP8266 goes into rst cause:2 after entering wifi credentials

### DISCUSSION THREAD

**Discussion by SimonFili** [arendst/Tasmota#19531] [Support request] [answered]
So I have this ESP8266 NodeMcu board that was working for a long time.  At one point, stopped working.

I'm suspecting a bad board, blown by static or bad power supply, anyhow, I want to explore with you guys if I can try something before filling my garbage can!

So I can flash properly with Tasmotizer, then I can connect to the web server properly, and set my Wi-Fi credential.  Then this happens in a loop.  I tried the light firmware, the normal one and the sensor one ( was this one before):

```
Using 'COM6' as serial port.
Showing logs:
[07:55:06]
[07:55:06]00:00:00.001 HDW: ESP8266EX
[07:55:06]00:00:00.029-035 CFG: Use defaults
[07:55:08]00:00:00.333 QPC: Count 1
[07:55:08]00:00:00.345 Project tasmota - Tasmota Version 13.1.0(lite)-2_7_4_9(2023-08-15T07:59:31)
[07:55:09]00:00:00.598 WIF: WifiManager active for 3 minutes
[07:55:09]00:00:00.108 HTP: Web server active on tasmota-D407A1-1953 with IP address 192.168.4.1
[07:55:15]00:00:07.100 QPC: Reset
[07:55:51]00:00:42.980 WIF: Connecting to AP Simon as tasmota-D407A1-1953 ...
[07:55:56]00:00:46.827 WIF: SSId Simon: Connected - IP Address 192.168.10.101
[07:56:05]00:00:56.429 APP: Restarting
[07:56:06]
[07:56:06] ets Jan  8 2013,rst cause:2, boot mode:(3,6)
[07:56:06]
[07:56:06]load 0x4010f000, len 3584, room 16 
[07:56:06]tail 0
[07:56:06]chksum 0xb0
[07:56:06]csum 0xb0
[07:56:06]v3969889e
[07:56:06]~ld
[07:56:06]
[07:56:06]00:00:00.001 HDW: ESP8266EX
[07:56:06]00:00:00.033-035 CFG: Use defaults
[07:56:08]00:00:00.331 QPC: Count 1
[07:56:08]00:00:00.342 Project tasmota - Tasmota Version 13.1.0(lite)-2_7_4_9(2023-08-15T07:59:31)
[07:56:08]00:00:00.596 WIF: WifiManager active for 3 minutes
[07:56:09]00:00:00.106 HTP: Web server active on tasmota-D407A1-1953 with IP address 192.168.4.1
[07:56:15]00:00:07.101 QPC: Reset
[08:06:07]00:09:59.046 APP: Serial logging disabled
```

Ideas to try?

**Comment by sfromis** [comment]
Restarting after entering Wifi credentials (`APP: Restarting`) is normal, but `CFG: Use defaults` is when Tasmota found no working config setup, and is doing initial config again. Could be a bad flash chip. Do make sure that your power source is not too weak.

**Comment by SimonFili** [comment]
Did a test while saving the Wi-Fi config:
I tested with many USB power supply, the best I have around gives me : 4.74V on the VCC side
On the 3.3v line, I captured a low value 3.24V (1ms capture using a Fluke)
 - So 0.06V below specs

Therefore, I don't think power supply voltage is the issue, but a bad flash ship could surely be.

Is there another way to test the flash?

**Reply by sfromis** [reply]
What I meant was not voltage (a deviation of 0.06V should be no problem), but power in milliamps. Anyway, measuring on "the 3.3V line" depends to a large extent on exactly where, in relation the decoupling caps, and even capacitance of wiring.

As for *ESP8266 flash write test*, that's a good Google search term, leading to something plausible-looking. Not tested it.

**Comment by SimonFili** [comment]
I've used the Arduino IDE and uploaded this sketch:
```
  ESP8266 CheckFlashConfig by Markus Sattler
  This sketch tests if the EEPROM settings of the IDE match to the Hardware
```

The test seems to work, but unsure if this is testing the flash?

```
Flash real id:   001640D8
Flash real size: 4194304 bytes

Flash ide  size: 4194304 bytes
Flash ide speed: 40000000 Hz
Flash ide mode:  DOUT
Flash Chip configuration ok.
```

I then modified the program to write to the first 7 memory of the EEPROM and read them.  All looks ok.

**Reply by Jason2866** [reply]
This sketch just tests the config of the flash. Not the flash chip itself. An easy way to find out replace the flash chip with a new one. If the issue is gone it is the flash. If not i would dump the module. Probably the esp8266 has died.

**Reply by SimonFili** [reply]
1) Does this code write to the flash (EEPROM), the last test I did I wrote to 512 memory location.
EEPROM.write(eepromStartAddress + i, i)

2) This is the module I have:
![image](https://github.com/arendst/Tasmota/assets/7669537/4f1afb0a-b2a9-4e89-b247-75545ec7a502)
I'm quite sure it's not easy to replace the flash.

3) I'm moving away from ESP8266 and moving everything to ESP32-S3, so, I was doing this for the fun of debugging (the best way to learn), but I will probably just scrap it if it's confirmed it's broken.

**Reply by SimonFili** [reply]
Here's the latest sketch used:

```
#include <EEPROM.h>

// EEPROM start address to write data
const int eepromStartAddress = 0;
// Data length
const int dataLength = 511;

void setup() {
  // Initialize serial and wait for port to open:
  Serial.begin(115200);
  // Wait for serial port to connect. Needed for native USB port only
  while (!Serial) {
    ; // wait for serial port to connect. Needed for native USB port only
  }

  // Initialize EEPROM
  EEPROM.begin(512);  // initialize with 512 bytes of EEPROM

  // Your existing code to check Flash configuration
  Serial.println("ESP8266 CheckFlashConfig");
  // ... (rest of your existing code)

  // Write data to EEPROM
  for (int i = 0; i < dataLength; i++) {
    EEPROM.write(eepromStartAddress + i, i);
  }
  EEPROM.commit();  // Commit changes to flash

  Serial.println("Written data to EEPROM.");

  // Read data from EEPROM
  Serial.println("Reading data from EEPROM:");
  for (int i = 0; i < dataLength; i++) {
    int readData = EEPROM.read(eepromStartAddress + i);
    Serial.print("Address: ");
    Serial.print(eepromStartAddress + i);
    Serial.print(" Data: ");
    Serial.println(readData);
  }

  // End EEPROM session
  EEPROM.end();
}

void loop() {
  // Nothing to do here
}

```

**Reply by sfromis** [reply]
> I'm quite sure it's not easy to replace the flash.

Just remove the metal shield. Inside of that, there is a SOP-8 flash chip  :nerd_face: Learning experience for you.....

**Reply by SimonFili** [reply]
Lol, well this is true, but I had my share of 10+ years of PCB / electronic design.  Thanks!

On the subject of the Flash/EEPROM
1) Am I right to say that FLASH = EEPROM here?
2) Is the above test a good one for testing this?

Thanks!

**Reply by sfromis** [reply]
Yes, flash=EEPROM. A "good" test is a charged question .... To me, a reasonable test would include ensuring that each and every bit can hold both 1 and 0, and that some "suitable" bit patterns survive unchanged. Less may still be able to find many defects, of course....

**Reply by arendst** [reply]
Arduino EEPROM is but a small area of total flash. It's just 1k bytes of total 4M!!

**Reply by SimonFili** [reply]
Alright, made some progress, I can confirm my chip info:
- Chip Type: ESP8266EX, confirming that your ESP8266MOD is based on the ESP8266EX chip.
- Flash Manufacturer: d8
- Flash Device: 4016
- Detected Flash Size: 4MB

@arendst you state the Arduino EEPROM is only 1k of the total 4M.

So, it's thus impossible to test the flash from Arduino code?

Thanks!

**Reply by SimonFili** [reply]
Alright, still playing around!

I did a small python script to create a test_blog.bin file:
```
with open("test_blob.bin", "wb") as f:
    for _ in range(4 * 1024 * 1024):  # 4MB
        f.write(b'\xAA')
```

Then used esptool to write the file to flash and got this:
```
C:\temp>esptool --port COM6 write_flash 0x00000 test_blob.bin
esptool.py v4.6.2
Serial port COM6
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: 3c:61:05:d4:07:a1
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x003fffff...
Compressed 4194304 bytes to 4087...
Wrote 4194304 bytes (4087 compressed) at 0x00000000 in 13.3 seconds (effective 2525.3 kbit/s)...
File  md5: 00296f79864928c317df2d2c0d5312a6
Flash md5: 996508570a80a996c57e0ea582e338b6
MD5 of 0xFF is 2b7a70fa59f8173635bcbe956bad56c6

A fatal error occurred: MD5 of file does not match data in flash!
```

Is this a valid test to prove the flash is BAD ?

**Reply by sfromis** [reply]
MD5 error does strongly suggest issues writing/reading flash, since the MD5 checksum does not match original. Assuming decent powering and wiring during flashing, the reasonable conclusion is bad flash. While there is no absolute guarantee of it not being the ESP8266 being wonky, I'd say that the remedy is to replace the flash chip, and more if you're "lazy".

**Reply by Jason2866** [reply]
Your test sketch seems working

```Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: f4:cf:a2:d4:40:7c
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x003fffff...
Compressed 4194304 bytes to 4087...
Wrote 4194304 bytes (4087 compressed) at 0x00000000 in 20.0 seconds (effective 1676.8 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

**Reply by Jason2866** [reply]
In my experience when flash erase (or write) time is much shorter than with an other device of same type and flash. Flash is defect.
Check how long a full `flash erase`takes

**Reply by SimonFili** [reply, marked as answer]
Alright, I found another NodeMCU like the failed one.  It's confirmed, the flash is fried!

Thus, I conclude this testing method is GOOD! 

**Good one:**
```
esptool --port COM6 write_flash 0x00000 test_blob.bin
esptool.py v4.6.2
Serial port COM6
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: e8:db:84:c5:9d:25
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x003fffff...
Compressed 4194304 bytes to 4087...
Wrote 4194304 bytes (4087 compressed) at 0x00000000 in 12.7 seconds (effective 2641.0 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

**BAD one**
```
esptool --port COM6 write_flash 0x00000 test_blob.bin
esptool.py v4.6.2
Serial port COM6
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: 3c:61:05:d4:07:a1
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x003fffff...
Compressed 4194304 bytes to 4087...
Wrote 4194304 bytes (4087 compressed) at 0x00000000 in 22.4 seconds (effective 1500.8 kbit/s)...
File  md5: 00296f79864928c317df2d2c0d5312a6
Flash md5: 4a35ed0f58ae77ca690b417175faf325
MD5 of 0xFF is 2b7a70fa59f8173635bcbe956bad56c6

A fatal error occurred: MD5 of file does not match data in flash!
```

@sfromis Since you seems to like playing around with soldering, I can send you my defective unit ;-)

**Reply by sfromis** [reply]
Not worth the postage :wink: And with most of my play time on the software side, I already have board fixing candidates if i somehow got in that mood!
